### PR TITLE
Datamodule

### PR DESCRIPTION
**Pull request by nateraw**
## What does this PR do?

Introduces `DataModule` as an _optional way_ to decouple data related hooks from `LightningModule`. Originally implemented in [`pytorch-lightning-bolts`](https://github.com/PyTorchLightning/pytorch-lightning-bolts), but moving to integrate with Lightning here.

Work in progress. We can play with the `DataModule` class a bunch more, but for now I'm looking for feedback on how they are handled in `trainer.py`.

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by nateraw** [comment]
Been playing with this script locally...including it here if anybody wants to try running it/breaking it.

```python
import os

import numpy as np
import torch
from torch import nn
from torch.nn import functional as F
from torch.utils.data import DataLoader, random_split
from torchvision.datasets import MNIST
from torchvision import datasets, transforms


import pytorch_lightning as pl


class CoolSystem(pl.LightningModule):
    
    def __init__(self, datamodule):
        super().__init__()
        self.datamodule = datamodule
        self.l1 = nn.Linear(28*28, 10)
        
    def forward(self, x):
        batch_size, channels, width, height = x.size()

        # (b, 1, 28, 28) -> (b, 1*28*28)
        x = x.view(batch_size, -1)

        return torch.log_softmax(self.l1(x), dim=1)

    def training_step(self, batch, batch_idx):
        x, y = batch
        logits = self(x)
        loss = F.nll_loss(logits, y)
        return {'loss': loss}

    def validation_step(self, batch, batch_idx):
        x, y = batch
        logits = self(x)
        loss = F.nll_loss(logits, y)
        return {'val_loss': loss}

    def validation_epoch_end(self, outputs):
        # called at the end of the validation epoch
        # outputs is an array with what you returned in validation_step for each batch
        # outputs = [{'loss': batch_0_loss}, {'loss': batch_1_loss}, ..., {'loss': batch_n_loss}] 
        avg_loss = torch.stack([x['val_loss'] for x in outputs]).mean()
        tensorboard_logs = {'val_loss': avg_loss}
        return {'avg_val_loss': avg_loss, 'log': tensorboard_logs, 'progress_bar': tensorboard_logs}

    def test_step(self, batch, batch_idx):
        x, y = batch
        logits = self(x)
        return {'pred': logits.detach().cpu().numpy()}
    
    def test_epoch_end(self, outputs):
        preds = np.concatenate([x["pred"] for x in outputs], axis=0)
        self.preds = np.argmax(preds, axis=1)

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
        return optimizer


class MNISTDataModule(pl.LightningDataModule):

    def __init__(self, data_dir=os.getcwd()):
        super().__init__()
        self.data_dir = data_dir

    def setup(self, stage):

        # transforms for images
        transform=transforms.Compose([
            transforms.ToTensor(),
            transforms.Normalize((0.1307,), (0.3081,))
        ])

        # prepare transforms standard to MNIST
        mnist_train = MNIST(self.data_dir, train=True, download=False, transform=transform)
        self.mnist_train, self.mnist_val = random_split(mnist_train, [55000, 5000])
        self.mnist_test = MNIST(self.data_dir, train=False, download=False, transform=transform)

    def prepare_data(self):

        # prepare transforms standard to MNIST
        MNIST(self.data_dir, train=True, download=True)
        MNIST(self.data_dir, train=False, download=True)

    def train_dataloader(self):
        return DataLoader(self.mnist_train, batch_size=32, num_workers=8)

    def val_dataloader(self):
        return DataLoader(self.mnist_val, batch_size=32, num_workers=8)
    
    def test_dataloader(self):
        return DataLoader(self.mnist_test, batch_size=32, num_workers=8)


if __name__ == '__main__':
    dm = MNISTDataModule()
    model = CoolSystem(dm)
    trainer = pl.Trainer(gpus=2, max_epochs=3, distributed_backend='ddp')
    # trainer = pl.Trainer(max_epochs=3)
    trainer.fit(model)
    trainer.test()
```

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -1109,6 +1112,24 @@ def __attach_dataloaders(self, model, train_dataloader=None, val_dataloaders=Non
...
             model.test_dataloader = _PatchDataLoader(test_dataloaders)
 
+    def __attach_datamodule(self, model, datamodule=None):
+        
+        # We use datamodule if it's been provided on .fit or .test, otherwise we check model for it
+        datamodule = datamodule or getattr(model, 'datamodule', None)
+
+        # If we have a datamodule, attach necessary hooks + dataloaders
+        if datamodule:
+            if self.is_overridden('setup', datamodule):
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:1122]: rather define this in `setup` of module
```
def setup(...):
    if getattr(self, 'datamodule', None):
        self.datamodule.setup(...)
```

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -1109,6 +1112,24 @@ def __attach_dataloaders(self, model, train_dataloader=None, val_dataloaders=Non
         if test_dataloaders is not None:
             model.test_dataloader = _PatchDataLoader(test_dataloaders)
 
+    def __attach_datamodule(self, model, datamodule=None):
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:1115]: how about the dataloader is runtime overwritten in previous `__attach_dataloaders`

File: pytorch_lightning/core/datamodule.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+            test_transforms=None,
+    ):
+        super().__init__()
+        self._train_transforms = train_transforms
+        self._val_transforms = val_transforms
+        self._test_transforms = test_transforms
+        self.dims = ()
+
+    @property
+    def train_transforms(self):
```
Comment (justusschock) [inline pytorch_lightning/core/datamodule.py:52]: Just a question: Why do you need these properties if neither setter nor getter are doing something specific like type checking? can't you just change this to public attributes?

Comment (nateraw) [reply pytorch_lightning/core/datamodule.py:52]: Going to add type checking there. does that make sense or should this still change to public attribute?

Comment (justusschock) [reply pytorch_lightning/core/datamodule.py:52]: No, with type checking it's fine :) Just in the current state it was not necessary to do it that way :)

Comment (awaelchli) [reply pytorch_lightning/core/datamodule.py:52]: one advantage of setters and getters is also that you can attach documentation to them, which you wouldn't be able to if it was just a regular attribute.

**Comment by williamFalcon** [comment]
@nateraw what if i don't pass in the dm to init, but instead do it in fit. Have we handled updating all the warnings etc?

**Comment by awaelchli** [comment]
I see a potential for a refactor here. It looks like the signature and docstrings for all these hooks like train_dataloader, val_dataloader etc. are going to be the same in LightningModule and DataModule. So if that's the case I propose to pull them out into a common interface like "DataHooks" and inherit it in LightningModule and DataModule. Otherwise, if the api changes we have to go and change docs and signature in two places :)

**Comment by nateraw** [comment]
> @nateraw what if i don't pass in the dm to init, but instead do it in fit. Have we handled updating all the warnings etc?

@williamFalcon Warnings aren't updated, but the following 4 cases should all work:

```python
dm = MNISTDataModule(**kwargs)
model = CoolSystem(datamodule=dm)  # self.datamodule = datamodule in __init__
```

```python
model = CoolSystem()
model.datamodule = dm
```

```python
model = CoolSystem()
trainer.fit(model, datamodule=dm)
```

```python
# Not sure why you'd want to do it this way, but you can.
model = CoolSystem()
trainer.fit(model)
trainer.test(model, datamodule)
```

If you pass a datamodule on .fit or .test, it will prefer that instead of one found at `model.datamodule`. Jirka mentioned adding misconfiguration exception when user trys to pass dataloaders + datamodule, which I think is a great idea too.

**Comment by nateraw** [comment]
> @nateraw I see a potential for a refactor here. It looks like the signature and docstrings for all these hooks like train_dataloader, val_dataloader etc. are going to be the same in LightningModule and DataModule. So if that's the case I propose to pull them out into a common interface like "DataHooks" and inherit it in LightningModule and DataModule. Otherwise, if the api changes we have to go and change docs and signature in two places :) What are your thoughts?

@awaelchli This is a wonderful idea. I'll throw that together. I didn't notice a `setup()` docstring anywhere...does it exist somewhere that I wasn't seeing?

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (nateraw) [reply pytorch_lightning/trainer/trainer.py]: Better yet, let's have DataModule do its own thing. Let's not mess with muddying up Trainer logic. 

We can encourage users to pass their subclass of `DataModule` directly to fit/test after running preparation on rank 0. Hopefully this makes things easier to manage 😄 . 

```python
if __name__ == '__main__':

    # Init your datamodule
    dm = MNISTDataModule()

    # Run all  preparation including downloading, transforms, etc.
    dm.prepare_data()

    # Get anything you need to init model, such as input dims
    input_dims = dm.size(1) * dm.size(2)

    # Init model
    model = CoolSystem(input_dims=input_dims)

    # Init Trainer
    trainer = pl.Trainer(max_epochs=3)

    # Train on your datamodule's data
    trainer.fit(model, datamodule=dm)
```

File: pytorch_lightning/core/datamodule.py (hunk 3 above)
Comment (nateraw) [reply pytorch_lightning/core/datamodule.py:63]: Strike my last point...when I viewed this earlier, it was on Mobile. I thought you were pointing out something else.

I'll add docs here as Adrian pointed out.

**Comment by nateraw** [comment]
Now it's like:

```python
if __name__ == '__main__':
    
    # Enforce random seed
    pl.seed_everything(42)

    # Init your datamodule
    # After subclass defined init, prepare_data and setup will run implicitly
    dm = MNISTDataModule()
    
    # Get anything from datamodule that you might need, such as input dims
    input_dims = np.prod(dm.size())

    # Init model
    model = CoolSystem(input_dims=input_dims)

    # Init Trainer
    trainer = pl.Trainer(max_epochs=1)

    # Train and test on your datamodule's data
    trainer.fit(model, datamodule=dm)
    trainer.test()
```

I think what we did with the DataModule wrapper is messing up `add_argparse_args`, though 🙁

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (Borda) [reply pytorch_lightning/trainer/trainer.py]: so you cannot pass datamodule to `CoolSystem`?

File: pytorch_lightning/core/datamodule.py [hunk 4]
```diff
@@ -9,12 +9,23 @@
...
         obj = type.__call__(cls, *args, **kwargs)
-        pd = rank_zero_only(obj.prepare_data)
-        obj.prepare_data = pd
+
+        # Wrap instance's prepare_data function with rank_zero_only and reassign to instance
+        obj.prepare_data = rank_zero_only(obj.prepare_data)
+
+        # Run both prepare_data() and setup() post-init
+        obj.prepare_data()
+        obj.setup()
```
Comment (awaelchli) [inline pytorch_lightning/core/datamodule.py:28]: I think we shouldn't call them manually here. 
These are hooks, so the trainer should call them, right?

Comment (nateraw) [reply pytorch_lightning/core/datamodule.py:28]: If we decouple from trainer, it allows datamodules to be used outside of Lightning (i.e. in pure Pytorch, if you'd like). I agree that moving out of implicit call after init is good idea though...just trying to keep this somewhat in line with what users are used to...any thoughts?

Comment (nateraw) [reply pytorch_lightning/core/datamodule.py:28]: Maybe `setup` calls `prepare_data` before running setup?

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (nateraw) [reply pytorch_lightning/trainer/trainer.py:1158]: Added conditionals to block this from happening 😄

**Comment by awaelchli** [comment]
@nateraw did the refactoring with an intermediate hooks class for train_dataloader etc not work out?

**Comment by nateraw** [comment]
@awaelchli I did it and it worked, but I think we want to think about that more. Decided to keep it simple and not rearrange a bunch of things for this PR 😄 .
